### ISSUE TITLE
Operation response schema with $ref showing as having no schema

### ISSUE THREAD

**Issue by mm-gmbd** [apigee-127/swagger-test-templates#92]
It looks like, based on Issue #16 and PR #27, that functionality was added to include assertions based on the operation response schemas. However, I realized that none of my tests have any such generated tests, but every test has the following code:

    res.body.should.equal(null); // non-json response or no schema

Every operation that I have defined in my Swagger has a response schema that defines a json response, so I was confused. Then I realized the common link between them all was that they are *all* defined in the Swagger root *responseDefinitions* object, and are all referenced with the appropriate $ref argument as defined in the Swagger spec. I'm guessing getting the schema from a reference has not been implemented?

Apologies if I am incorrect!

Thanks,
-Max

**Comment by noahdietz** [comment]
@mm-gmbd you are correct. The Javascript API of this module does not resolve definition references. 

This tool was designed in parallel and for main use with the [swagger-node](https://www.npmjs.com/package/swagger) module. There, we implemented a CLI that actually resolved the references prior to test generation. 

As such, the reference resolution here still needs to be implemented :)

**Comment by mm-gmbd** [comment]
@noahdietz, thanks for the information. I've done some work on a fork of [swagger-js-codegen](https://github.com/wcandillon/swagger-js-codegen) that resolves references, but if there is written/accepted code that resolves these references in [swagger-node](https://github.com/swagger-api/swagger-node) that you could point me to, I can make another fork to implement this.

I'll take a peek at swagger-node, but I thought you could probably point me in the right direction :) thanks in advance!

**Comment by mm-gmbd** [comment]
After some quick inspection, I didn't find anything that looked like [swagger-node](https://github.com/swagger-api/swagger-node) that looked like there was reference support. Searching "ref" in the repository resulted in a few results, but they were all in markdown files and swagger YAML definitions, not actual code resolving references.

However, it looks like the chain is that *swagger-node* uses [swagger-tools](https://github.com/apigee-127/swagger-tools), and *swagger-tools* uses [json-refs](https://github.com/whitlockjc/json-refs). 

So, would the preferred strategy be to add *json-refs* (rather than roll my own) and use the resolveRefs or resolveLocalRefs (as defined in the [API docs](https://github.com/whitlockjc/json-refs/blob/master/docs/API.md))?

**Comment by noahdietz** [comment]
@mm-gmbd I should've been more specific, I apologize. There is functionality to resolve references in [swagger-tools](https://github.com/apigee-127/swagger-tools/blob/master/lib/specs.js#L651). I was referring to the [use of this](https://github.com/swagger-api/swagger-node/blob/master/lib/commands/project/project.js#L433) in swagger-node prior to any test generation. There is no logging that states we are resolving references for the user, it was a behind the scenes thing.

**Comment by mm-gmbd** [comment]
Hmm. So if swagger-tools is used, and swagger-tools has support for references, shouldn't swagger-test-templates already resolve those references correctly?

Does swagger-test-templates just need to use the latest swagger-tools to fix? I see that there is [an issue in swagger-tools for supporting external references](https://github.com/apigee-127/swagger-tools/issues/227), but these are internal definitions, so I would think that wouldn't be an issue.

Apologies for my confusion. I'm just trying to figure out what needs to be done for this to work :)

**Comment by noahdietz** [comment]
Swagger-test-templates does not resolve any references. It would be great to have that added. 

What needs to happen is that the swagger given to test-templates needs to be run through the swagger-tools reference resolution function before being processed.

**Comment by theganyo** [comment]
(Actually, all of this should be ported to use Sway instead of swagger-tools.)

**Comment by noahdietz** [comment]
This project doesn't even have a swagger-tools dependency, so if reference resolution is added, using sway is breaking new ground. Good timing, should be a quick addition

**Comment by theganyo** [comment]
Ah, good! Well, Swagger-Node should be ported. ;)

**Comment by noahdietz** [comment]
@mm-gmbd, are you planning on opening a PR for this functionality? I'm planning on spending some time here and a I want to prioritize. Thanks!

**Comment by mm-gmbd** [comment]
@noahdietz, apologies for the radio silence, just getting back into things from the holidays. At first, I was thinking this was a quick fix (i.e. add a call to `.resolve()`) or something like that, but I'm guessing porting to Sway from swagger-tools would be a bit more? I'd like to get this resolved quickly, so I may take a stab at it. I'll have a better idea in the next day or so.

**Comment by noahdietz** [comment]
@mm-gmbd you're right, it won't be as easy as ```.resolve()``` but its not too much harder I believe. I dug through the Sway repo. It looks as though you must create a new SwaggerApi instance by using ```.createSwaggerApi()``` and passing the Swagger definition being used. The SwaggerApi object will have a ```resolved``` property that is the fully resolved swagger definition. This is the one that should be fed into testGen() I believe. It seems as though it is a JSON object that will behave with this module.

This will also be a decent catch for a bad Swagger with the validation Sway does on creation of new objects.

**Comment by ferni** [comment]
Hi, may I suggest a pragmatic approach? Currently the code generator generates something like this in the middle of a test:

```
var schema = {
  "type": "object",
  "properties": {
    "pagination": {
      "$ref": "#/definitions/Pagination"
    }
  }
}
```

What if we modify the test generator to add the definitions object, like this:

```
var swagger = require('<path to swagger.json>');

var schema = {
  "type": "object",
  "properties": {
    "pagination": {
      "$ref": "#/definitions/Pagination"
    }
  },
  "definitions": swagger.definitions
}
```

**Comment by ferni** [comment]
To solve this I just used this simple tool: https://www.npmjs.com/package/json-schema-deref

Just "deref" it before passing it to the test generator.

**Comment by noahdietz** [comment]
@ferni are you using the asynchronous module? It would be trivial to add the [synchronous version](https://www.npmjs.com/package/json-schema-deref-sync) of this module, combined with [js-yaml](https://www.npmjs.com/package/js-yaml) so we can take in a swagger.yaml file as input always.

**Comment by ferni** [comment]
I used the async one, but I don't think it would make much of a difference if I used the synchronous version.

**Comment by noahdietz** [comment]
@ferni ok, it would be really cool if you could put together a PR that utilizes the synchronous version within testGen to deref the JSON. If you wanted to go so far as to add support for reading in a yaml file prior to deref, that would be much appreciated.

**Comment by ferni** [comment]
Ok, I'll do that tonight. Should the code detect if the file is YAML or JSON, and in the case it is YAML, convert it to JSON? Or should it be assumed that it is always YAML?

**Comment by noahdietz** [comment]
the former option would be preferred. Or you can add an option whose presence means the input is a yaml file. whichever you'd prefer. But assuming yaml would be tough because all of the tests use a swagger.json input file :)

**Comment by ferni** [comment]
I did the defer thing, I didn't do the YAML thing yet, but one thing at a time is just as well right?

**Comment by mm-gmbd** [comment]
Should this be considered closed with PR #103? Thanks @ferni!
